File: api-gateway/src/main/java/com/jmsoftware/maf/apigateway/security/impl/RbacReactiveAuthorizationManagerImpl.java
package com.jmsoftware.maf.apigateway.security.impl;

import cn.hutool.core.util.StrUtil;
import com.google.common.collect.Lists;
import com.jmsoftware.maf.apigateway.remoteapi.AuthCenterRemoteApi;
import com.jmsoftware.maf.common.constant.MafHttpHeader;
import com.jmsoftware.maf.common.domain.authcenter.permission.GetPermissionListByRoleIdListPayload;
import com.jmsoftware.maf.common.domain.authcenter.permission.GetPermissionListByRoleIdListResponse;
import com.jmsoftware.maf.common.domain.authcenter.permission.PermissionType;
import com.jmsoftware.maf.common.domain.authcenter.role.GetRoleListByUserIdSingleResponse;
import com.jmsoftware.maf.common.domain.authcenter.security.UserPrincipal;
import com.jmsoftware.maf.common.exception.SecurityException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.springframework.http.HttpStatus;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.security.authorization.AuthorizationDecision;
import org.springframework.security.authorization.ReactiveAuthorizationManager;
import org.springframework.security.core.Authentication;
import org.springframework.security.web.server.authorization.AuthorizationContext;
import org.springframework.stereotype.Component;
import org.springframework.util.AntPathMatcher;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * Description: RbacReactiveAuthorizationManagerImpl
 * <p>
 * Implementation of RBAC (Role-based access control) reactive authorization manager
 *
 * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com, date: 12/29/2020 9:54 AM
 * @see <a href='https://en.wikipedia.org/wiki/Role-based_access_control'>Role-based access control</a>
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class RbacReactiveAuthorizationManagerImpl implements ReactiveAuthorizationManager<AuthorizationContext> {
    private final AntPathMatcher antPathMatcher = new AntPathMatcher();
    private final AuthCenterRemoteApi authCenterRemoteApi;

    /**
     * Retrieve roles flux.
     *
     * @param userPrincipalMono the user principal mono
     * @return the flux
     */
    private Flux<GetRoleListByUserIdSingleResponse> retrieveRoles(Mono<UserPrincipal> userPrincipalMono) {
        // Get role list by user ID, and then convert to Flux<?>
        return userPrincipalMono
                .flatMap(userPrincipal -> this.authCenterRemoteApi.getRoleListByUserId(userPrincipal.getId()))
                .flatMapMany(Flux::fromIterable)
                .switchIfEmpty(Flux.error(new SecurityException(HttpStatus.UNAUTHORIZED, "Roles not assigned!")));
    }

    /**
     * Filter roles mono.
     *
     * @param roleFlux the role flux
     * @return the mono
     */
    private Mono<List<Long>> mapRole(Flux<GetRoleListByUserIdSingleResponse> roleFlux) {
        return roleFlux
                .map(GetRoleListByUserIdSingleResponse::getId)
                .collectList()
                .switchIfEmpty(roleFlux.map(GetRoleListByUserIdSingleResponse::getId).collectList());
    }

    /**
     * Retrieve permissions mono.
     *
     * @param roleIdListMono the role id list mono
     * @return the mono
     */
    private Mono<List<GetPermissionListByRoleIdListResponse.Permission>> retrievePermissions(Mono<List<Long>> roleIdListMono) {
        // Get permission list based on the Mono<List<Long>>
        // auth-center will respond /** for role "admin"
        return roleIdListMono.flatMap(
                roleIdList -> {
                    val payload = new GetPermissionListByRoleIdListPayload();
                    payload.setRoleIdList(roleIdList);
                    payload.setPermissionTypeList(Lists.newArrayList(PermissionType.BUTTON));
                    return this.authCenterRemoteApi.getPermissionListByRoleIdList(payload);
                })
                .switchIfEmpty(Mono.error(new SecurityException(HttpStatus.FORBIDDEN, "Permission not found!")));
    }

    @Override
    public Mono<AuthorizationDecision> check(Mono<Authentication> authentication, AuthorizationContext object) {
        val request = object.getExchange().getRequest();
        val userPrincipalMono = authentication.map(auth -> (UserPrincipal) auth.getPrincipal());
        val roleFlux = this.retrieveRoles(userPrincipalMono);
        val roleIdListMono = this.mapRole(roleFlux);
        val permissionListMono = this.retrievePermissions(roleIdListMono);
        // Aggregate 2 Mono
        val zip = Mono.zip(permissionListMono, userPrincipalMono);
        return zip.map(mapper -> {
            val permissionList = mapper.getT1();
            val buttonPermissionList = permissionList.stream()
                    .filter(permission -> StrUtil.isNotBlank(permission.getUrl()))
                    .filter(permission -> StrUtil.isNotBlank(permission.getMethod()))
                    .collect(Collectors.toList());
            val userPrincipal = mapper.getT2();
            for (var buttonPermission : buttonPermissionList) {
                if (this.checkRestfulAccess(buttonPermission, request)) {
                    log.info("Authorization success! Resource [{}] {} is accessible for user(username: {})",
                             request.getMethod(), request.getURI(), userPrincipal.getUsername());
                    request
                            .mutate()
                            .headers(httpHeaders -> httpHeaders.set(MafHttpHeader.X_USERNAME.getHeader(),
                                                                    userPrincipal.getUsername()))
                            .build();
                    return new AuthorizationDecision(true);
                }
            }
            log.warn("Authorization failure! Resource [{}] {} is not accessible for user(username: {})",
                     request.getMethod(), request.getURI(), userPrincipal.getUsername());
            return new AuthorizationDecision(false);
        });
    }

    /**
     * <p>Check Restful access.</p>
     * <ul>
     * <li>Check if the URL is matched</li>
     * <li>Check if the HTTP method is matched</li>
     * </ul>
     *
     * @param buttonPermission the button permission
     * @param request          the request
     * @return true: accessible; false: not accessible
     * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com, date: 1/13/2021 11:04 AM
     */
    private boolean checkRestfulAccess(GetPermissionListByRoleIdListResponse.Permission buttonPermission,
                                       ServerHttpRequest request) {
        val urlMatched = this.antPathMatcher.match(buttonPermission.getUrl(), request.getURI().getPath());
        // "*" is for super user. Super user's permission is like URL: "/**", method: "*"
        val allMethods = StrUtil.equals(buttonPermission.getMethod(), "*");
        if (allMethods) {
            return urlMatched;
        }
        val methodMatched = Objects.requireNonNull(request.getMethod()).matches(buttonPermission.getMethod());
        return urlMatched && methodMatched;
    }
}

File: api-gateway/api-gateway-biz/src/main/kotlin/com/jmsoftware/maf/apigateway/security/impl/RbacReactiveAuthorizationManagerImpl.kt
package com.jmsoftware.maf.apigateway.security.impl

import cn.hutool.core.util.StrUtil
import com.jmsoftware.maf.apigateway.remote.AuthCenterWebClientService
import com.jmsoftware.maf.common.constant.MafHttpHeader.X_ID
import com.jmsoftware.maf.common.constant.MafHttpHeader.X_USERNAME
import com.jmsoftware.maf.common.domain.authcenter.permission.GetPermissionListByRoleIdListPayload
import com.jmsoftware.maf.common.domain.authcenter.permission.Permission
import com.jmsoftware.maf.common.domain.authcenter.permission.PermissionType
import com.jmsoftware.maf.common.domain.authcenter.role.GetRoleListByUserIdSingleResponse
import com.jmsoftware.maf.common.domain.authcenter.security.UserPrincipal
import com.jmsoftware.maf.common.exception.SecurityException
import com.jmsoftware.maf.common.util.logger
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus.FORBIDDEN
import org.springframework.http.HttpStatus.UNAUTHORIZED
import org.springframework.http.server.reactive.ServerHttpRequest
import org.springframework.security.authorization.AuthorizationDecision
import org.springframework.security.authorization.ReactiveAuthorizationManager
import org.springframework.security.core.Authentication
import org.springframework.security.web.server.authorization.AuthorizationContext
import org.springframework.util.AntPathMatcher
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.util.function.Tuple2

/**
 * # RbacReactiveAuthorizationManagerImpl
 *
 * Implementation of RBAC (Role-based access control) reactive authorization manager
 *
 * @author Johnny Miller (锺俊), e-mail: johnnysviva@outlook.com, date: 4/16/22 8:57 PM
 * @see [Role-based access control](https://en.wikipedia.org/wiki/Role-based_access_control)
 */
open class RbacReactiveAuthorizationManagerImpl(
    private val authCenterWebClientService: AuthCenterWebClientService
) : ReactiveAuthorizationManager<AuthorizationContext> {
    companion object {
        private val log = logger()
    }

    private val antPathMatcher = AntPathMatcher()

    /**
     * Retrieve roles flux.
     *
     * @param userPrincipalMono the user principal mono
     * @return the flux
     */
    private fun retrieveRoles(userPrincipalMono: Mono<UserPrincipal>): Flux<GetRoleListByUserIdSingleResponse> {
        // Get role list by user ID, and then convert to Flux<?>
        return userPrincipalMono
            .flatMap { userPrincipal: UserPrincipal -> authCenterWebClientService.getRoleListByUserId(userPrincipal.id!!) }
            .flatMapMany { Flux.fromIterable(it) }
            .switchIfEmpty(Flux.error(SecurityException("Roles not assigned!", UNAUTHORIZED)))
    }

    /**
     * Filter roles mono.
     *
     * @param roleFlux the role flux
     * @return the mono
     */
    private fun mapRole(roleFlux: Flux<GetRoleListByUserIdSingleResponse>): Mono<List<Long>> {
        return roleFlux
            .map(GetRoleListByUserIdSingleResponse::id)
            .collectList()
            .switchIfEmpty(roleFlux.map(GetRoleListByUserIdSingleResponse::id).collectList())
    }

    /**
     * Retrieve permissions mono.
     *
     * @param roleIdListMono the role id list mono
     * @return the mono
     */
    private fun retrievePermissions(roleIdListMono: Mono<List<Long>>): Mono<List<Permission>> {
        // Get permission list based on the Mono<List<Long>>
        // auth-center will respond /** for role "admin"
        return roleIdListMono.flatMap { roleIdList: List<Long> ->
            val payload = GetPermissionListByRoleIdListPayload()
            payload.roleIdList = roleIdList
            payload.permissionTypeList = listOf(PermissionType.BUTTON)
            authCenterWebClientService.getPermissionListByRoleIdList(payload)
        }
            .switchIfEmpty(Mono.error(SecurityException("Permission not found!", FORBIDDEN)))
    }

    override fun check(
        authentication: Mono<Authentication>,
        `object`: AuthorizationContext
    ): Mono<AuthorizationDecision> {
        val request = `object`.exchange.request
        val userPrincipalMono = authentication.map { auth: Authentication -> auth.principal as UserPrincipal }
        val roleFlux = retrieveRoles(userPrincipalMono)
        val roleIdListMono = mapRole(roleFlux)
        val permissionListMono = retrievePermissions(roleIdListMono)
        // Aggregate 2 Mono
        val zip = Mono.zip(permissionListMono, userPrincipalMono)
        return zip.map { mapper: Tuple2<List<Permission>, UserPrincipal> ->
            val userPrincipal = mapper.t2
            val anyMatched = mapper.t1.stream()
                .filter { permission: Permission -> StrUtil.isNotBlank(permission.url) }
                .filter { permission: Permission -> StrUtil.isNotBlank(permission.method) }
                .anyMatch { permission: Permission ->
                    checkRestfulAccess(permission, request)
                }
            if (anyMatched) {
                log.info("Authorization success! Resource [${request.method}] ${request.uri} is accessible for user(username: ${userPrincipal.username})")
                request.mutate()
                    .headers { httpHeaders: HttpHeaders ->
                        httpHeaders[X_ID.header] = userPrincipal.id.toString()
                        httpHeaders[X_USERNAME.header] = userPrincipal.username
                    }.build()
                AuthorizationDecision(true)
            } else {
                log.warn("Authorization failure! Resource [${request.method}] ${request.uri} is not accessible for user(username: ${userPrincipal.username})")
                AuthorizationDecision(false)
            }
        }
    }

    /**
     *
     * Check Restful access.
     *
     *  * Check if the URL is matched
     *  * Check if the HTTP method is matched
     *
     *
     * @param buttonPermission the button permission
     * @param request          the request
     * @return true: accessible; false: not accessible
     * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com, date: 1/13/2021 11:04 AM
     */
    private fun checkRestfulAccess(
        buttonPermission: Permission,
        request: ServerHttpRequest
    ): Boolean {
        val urlMatched = antPathMatcher.match(buttonPermission.url!!, request.uri.path)
        // "*" is for super user. Super user's permission is like URL: "/**", method: "*"
        val allMethods = StrUtil.equals(buttonPermission.method, "*")
        if (allMethods) {
            return urlMatched
        }
        val methodMatched = request.method!!.matches(buttonPermission.method!!)
        return urlMatched && methodMatched
    }
}
